boards/shared/Src/Io/Io_SharedSpi.c

```diff
@@ -0,0 +1,117 @@
+#include <assert.h>
+#include <stdlib.h>
+#include "Io_SharedSpi.h"
+
+#define EXIT_IF_STATUS_NOT_OK(status, nss_port, nss_pin)          \
+    if ((status) != HAL_OK)                                       \
+    {                                                             \
+        HAL_GPIO_WritePin((nss_port), (nss_pin), GPIO_PIN_RESET); \
+        return (status);                                          \
+    }
+
+struct SharedSpi
+{
+    SPI_HandleTypeDef *hspi;
+    GPIO_TypeDef *     nss_port;
+    uint16_t           nss_pin;
+};
+
+struct SharedSpi *Io_SharedSpi_Create(
+    SPI_HandleTypeDef *hspi,
+    GPIO_TypeDef *     nss_port,
+    uint16_t           nss_pin)
+{
+    assert(hspi != NULL);
+
+    struct SharedSpi *spi = malloc(sizeof(struct SharedSpi));
+    assert(spi != NULL);
+
+    spi->hspi     = hspi;
+    spi->nss_pin  = nss_pin;
+    spi->nss_port = nss_port;
+
+    return spi;
+}
+
+void Io_SharedSpi_SetNssLow(const struct SharedSpi *const spi)
+{
+    HAL_GPIO_WritePin(spi->nss_port, spi->nss_pin, GPIO_PIN_RESET);
+}
+
+void Io_SharedSpi_SetNssHigh(const struct SharedSpi *const spi)
+{
+    HAL_GPIO_WritePin(spi->nss_port, spi->nss_pin, GPIO_PIN_SET);
+}
+
+HAL_StatusTypeDef Io_SharedSpi_TransmitAndReceive(
+    const struct SharedSpi *const spi,
+    uint8_t *                     tx_data,
+    uint16_t                      tx_size,
+    uint8_t *                     rx_data,
+    uint16_t                      rx_size)
+
+{
+    Io_SharedSpi_SetNssLow(spi);
+    EXIT_IF_STATUS_NOT_OK(
+        HAL_SPI_Transmit(spi->hspi, tx_data, tx_size, 100U), spi->nss_port,
+        spi->nss_pin)
+    EXIT_IF_STATUS_NOT_OK(
+        HAL_SPI_Receive(spi->hspi, rx_data, rx_size, 100U), spi->nss_port,
+        spi->nss_pin)
+    Io_SharedSpi_SetNssHigh(spi);
+
+    return HAL_OK;
+}
+
+HAL_StatusTypeDef Io_SharedSpi_Transmit(
+    const struct SharedSpi *const spi,
+    uint8_t *                     tx_data,
+    uint16_t                      tx_size)
+{
+    Io_SharedSpi_SetNssLow(spi);
+    HAL_StatusTypeDef status =
+        HAL_SPI_Transmit(spi->hspi, tx_data, tx_size, 100U);
+    Io_SharedSpi_SetNssHigh(spi);
+
+    return status;
+}
+
+HAL_StatusTypeDef Io_SharedSpi_Receive(
+    const struct SharedSpi *const spi,
+    uint8_t *                     rx_data,
+    uint16_t                      rx_size)
+{
+    Io_SharedSpi_SetNssLow(spi);
+    HAL_StatusTypeDef status =
+        HAL_SPI_Receive(spi->hspi, rx_data, rx_size, 100U);
+    Io_SharedSpi_SetNssHigh(spi);
+
+    return status;
+}
+
+HAL_StatusTypeDef Io_SharedSpi_MultipleTransmitWithoutNssToggle(
+    const struct SharedSpi *const spi,
+    uint8_t *                     tx_data,
+    uint16_t                      tx_size,
+    size_t                        num_tx_data_copies)
+{
+    HAL_StatusTypeDef status;
+    for (size_t i = 0U; i < num_tx_data_copies; i++)
+    {
+        status = HAL_SPI_Transmit(spi->hspi, tx_data, tx_size, 100U);
```

Comment:
what is `100U` here? Is that a timeout? I would try to avoid doing operations like this synchronously, as if this started timing out, even just a few calls to this function could run over the time limit for an RTOS task.

Reply:
Okay that makes sense. Were you thinking of moving away from blocking functions in general? Or were you thinking of just leaving the timeout duration as 0? @garethellis0

Reply:
I was testing HAL_SPI_Transmit_DMA. For the number of packages that I was dealing this it didn't help much in terms of performance. Also ran into this [link](https://community.st.com/s/question/0D50X00009sW5Ld/halspitransmitit-blocks) cuz I ran into a similar problem with HAL_SPI_DMA.